### PR TITLE
Add metadata to sweeps

### PR DESCRIPTION
**Pull request by dstrain115**
- Adds optional metadata attribute to Linspace and Points.
- This will be used to adjust device parameters via sweep.

See https://tinyurl.com/cirq-sweep-metadata-public for RFC.

### CODE REVIEW

File: cirq-core/cirq/study/sweeps.py [hunk 1]
```diff
@@ -470,12 +504,17 @@ def _values(self) -> Iterator[float]:
                 yield self.start * (1 - p) + self.stop * p
 
     def __repr__(self) -> str:
+        metadata_repr = f', metadata={self.metadata!r}' if self.metadata else ""
```
Comment (maffoo) [inline cirq-core/cirq/study/sweeps.py:507]: here too:
```suggestion
        metadata_repr = f', metadata={self.metadata!r}' if self.metadata is not None else ""
```

**Comment by maffoo** [comment]
Should probably add metadata on `ListSweep` as well, since that also can be serialized and sent to the API (https://github.com/quantumlib/Cirq/blob/master/cirq-google/cirq_google/api/v2/sweeps.py#L60).

**Comment by maffoo** [comment]
Note that in the `ListSweep` case, the sweep is converted during serialization into a zip over individual `Point` sweeps when serializing, so we probably want the `ListSweep` metadata to be something like `Mapping[cirq.TParamKey, Any]` so that the metadata can be split up appropriately. Similarly in the "flatten expressions" transformer we will want to be able to collect metadata about the individual param keys into a map and attach it to the `ListSweep`s that are produced.

**Comment by dstrain115** [comment]
Ok, thanks for the corrections.  I will save ListSweep for a follow-up PR.  I want to give it a bit more though.

**Pull request merged**